types/basetypes/dynamic_type.go

```diff
@@ -0,0 +1,165 @@
+	if !in.IsKnown() {
+		return NewDynamicUnknown(), nil
+	}
+
+	if in.IsNull() {
+		return NewDynamicNull(), nil
+	}
+
+	// MAINTAINER NOTE: It should not be possible for Terraform core to send a known value of `tftypes.DynamicPseudoType`.
+	// This check prevents an infinite recursion that would result if this scenario occurs when attempting to create a dynamic value.
+	if in.Type().Is(tftypes.DynamicPseudoType) {
+		return nil, errors.New("ambiguous known value for `tftypes.DynamicPseudoType` detected")
+	}
+
+	attrType := t.DetermineAttrType(in.Type())
+	val, err := attrType.ValueFromTerraform(ctx, in)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewDynamicValue(val), nil
+}
+
+// ValueType returns the Value type.
+func (t DynamicType) ValueType(_ context.Context) attr.Value {
+	return DynamicValue{}
+}
+
+// DetermineAttrType returns the appropriate attr.Type equivalent for a given tftypes.Type
+func (t DynamicType) DetermineAttrType(in tftypes.Type) attr.Type {
+	// Primitive types
+	if in.Is(tftypes.Bool) {
+		return BoolType{}
+	}
+	if in.Is(tftypes.Number) {
+		return NumberType{}
+	}
+	if in.Is(tftypes.String) {
+		return StringType{}
+	}
+	if in.Is(tftypes.DynamicPseudoType) {
+		// Null and Unknown values that do not have a type determined will have a type of DynamicPseudoType
+		return DynamicType{}
+	}
+
+	// Collection types
+	if in.Is(tftypes.List{}) {
+		//nolint:forcetypeassert // Type assertion is guaranteed by the above `(tftypes.Type).Is` function
+		l := in.(tftypes.List)
+
+		elemType := t.DetermineAttrType(l.ElementType)
+		return ListType{ElemType: elemType}
+	}
+	if in.Is(tftypes.Map{}) {
+		//nolint:forcetypeassert // Type assertion is guaranteed by the above `(tftypes.Type).Is` function
+		m := in.(tftypes.Map)
+
+		elemType := t.DetermineAttrType(m.ElementType)
+		return MapType{ElemType: elemType}
+	}
+	if in.Is(tftypes.Set{}) {
+		//nolint:forcetypeassert // Type assertion is guaranteed by the above `(tftypes.Type).Is` function
+		s := in.(tftypes.Set)
+
+		elemType := t.DetermineAttrType(s.ElementType)
+		return SetType{ElemType: elemType}
+	}
+
+	// Structural types
+	if in.Is(tftypes.Object{}) {
+		//nolint:forcetypeassert // Type assertion is guaranteed by the above `(tftypes.Type).Is` function
+		o := in.(tftypes.Object)
+
+		attrTypes := make(map[string]attr.Type, len(o.AttributeTypes))
+		for name, tfType := range o.AttributeTypes {
+			attrTypes[name] = t.DetermineAttrType(tfType)
+		}
+		return ObjectType{AttrTypes: attrTypes}
+	}
+	if in.Is(tftypes.Tuple{}) {
+		//nolint:forcetypeassert // Type assertion is guaranteed by the above `(tftypes.Type).Is` function
+		tup := in.(tftypes.Tuple)
+
+		elemTypes := make([]attr.Type, len(tup.ElementTypes))
+		for i, tfType := range tup.ElementTypes {
+			elemTypes[i] = t.DetermineAttrType(tfType)
+		}
+		return TupleType{ElemTypes: elemTypes}
+	}
+
+	panic(fmt.Sprintf("unsupported tftypes.Type detected: %T", in))
```

Comment:
Would like some feedback here, I wasn't sure how best to surface this type of scenario. Technically impossible, so I originally didn't introduce an `error` into the `(attr.TypeWithDynamicValue).DetermineAttrType` interface, however I'd love to avoid panics like this 🙁

Reply:
+1 to introducing a second `error` return for this, especially because there are not too many callers 👍

Reply:
This function has been adjusted to be an internal helper to the `types.DynamicType` struct. Replaced this panic with an error c920b8a ⚡